reactos/drivers/storage/storahci/storahci.c

```diff
@@ -768,53 +1189,65 @@ IsPortValid (
 }// -- IsPortValid()
 
 /**
- * @name DeviceInquiryRequest
+ * @name AddQueue
  * @implemented
  *
- * Tells wheather given port is implemented or not
+ * Add Srb to Queue
  *
- * @param AdapterExtension
+ * @param Queue
  * @param Srb
- * @param Cdb
  *
  * @return
- * return STOR status for DeviceInquiryRequest
+ * return TRUE if Srb is successfully added to Queue
  *
- * @remark
- * http://www.seagate.com/staticfiles/support/disc/manuals/Interface%20manuals/100293068c.pdf
  */
-ULONG
-DeviceInquiryRequest (
-    __in PAHCI_ADAPTER_EXTENSION AdapterExtension,
-    __in PSCSI_REQUEST_BLOCK Srb,
-    __in PCDB Cdb
+__inline
+BOOLEAN
+AddQueue (
+    __inout PAHCI_QUEUE Queue,
+    __in PVOID Srb
     )
 {
-    PVOID DataBuffer;
-    ULONG DataBufferLength;
+    NT_ASSERT(Queue->Head < MAXIMUM_QUEUE_BUFFER_SIZE);
+    NT_ASSERT(Queue->Tail < MAXIMUM_QUEUE_BUFFER_SIZE);
 
-    DebugPrint("DeviceInquiryRequest()\n");
+    if (Queue->Head == ((Queue->Tail + 1) % MAXIMUM_QUEUE_BUFFER_SIZE))

```

Comment:
Don't you mean if Head + 1 == Tail?

Head=0, Tail=0 -> empty queue
Head=1, Tail=0 -> 1 item
Head=253, Tail=0 -> 253 items
Head=254, Tail=0 -> 254 items; add one more to Head and the queue overflows, this is is the one that should fail.